File: xdm-core/src/main/java/xdm/core/downloaders/AbstractDownloader.java
package xdm.core.downloaders;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Comparator;
import java.util.Date;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.Stream;
import lombok.Getter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import xdm.core.Config;
import xdm.core.DownloadProgressListener;
import xdm.core.InteractiveCredentialProvider;
import xdm.core.util.HttpDateParser;

public abstract class AbstractDownloader {
  private final Logger logger = LoggerFactory.getLogger(AbstractDownloader.class);
  @Getter protected final DownloaderType downloaderType;
  protected AtomicBoolean stopFlag = new AtomicBoolean(false);
  protected AtomicLong length = new AtomicLong();
  protected String folder;
  @Getter protected final long id;
  protected AtomicBoolean finished = new AtomicBoolean(false);
  @Getter protected int maxCount = 8;
  protected final DownloadProgressListener listener;
  protected SpeedLimiter speedLimiter;
  protected AtomicLong downloaded = new AtomicLong(0);
  @Getter protected AtomicInteger progress = new AtomicInteger(0);
  protected long lastUpdated;
  protected long lastSaved;
  @Getter protected boolean assembling;
  @Getter protected float downloadSpeed;
  @Getter protected long eta;
  protected int outputFormat;
  @Getter protected boolean converting;
  protected int convertPrg;
  protected String lastModified;
  protected InteractiveCredentialProvider credentialProvider;

  public abstract void start();

  public abstract void stop();

  public abstract void resume();

  public abstract int getType();

  public long getSize() {
    return length.get();
  }

  public long getDownloaded() {
    return downloaded.get();
  }

  public abstract Metadata getMetadata();

  public abstract SegmentDetails getSegmentDetails();

  public void setOutputMediaFormat(int format) {
    this.outputFormat = format;
  }

  protected AbstractDownloader(
      long id,
      DownloaderType downloaderType,
      DownloadProgressListener listener,
      InteractiveCredentialProvider credentialProvider) {
    this.id = id;
    this.downloaderType = downloaderType;
    this.speedLimiter = new SpeedLimiter(this);
    this.listener = listener;
    this.credentialProvider = credentialProvider;
  }

  public synchronized void cleanup() {
    logger.info("Cleaning up temp files");
    try {
      Path folderPath = Paths.get(folder);
      if (!Files.exists(folderPath)) {
        return;
      }
      try (Stream<Path> files = Files.walk(folderPath).sorted(Comparator.reverseOrder())) {
        files.forEach(
            f -> {
              try {
                Files.delete(f);
                logger.info("Successfully delete file: {}", f);
              } catch (IOException e) {
                logger.error("Error deleting temp file: {}", f, e);
              }
            });
      }
    } catch (Exception ex) {
      logger.error("Error cleaning up temp files", ex);
    }
  }

  protected String getOutputFolder() {
    return listener.getOutputFolder(id);
  }

  protected File getBackupFile(String folder) {
    File f = new File(folder);
    File[] files = f.listFiles();
    if (files == null || files.length < 1) return null;
    for (File file : files) {
      if (file.getName().endsWith(".bak")) {
        return file;
      }
    }
    return null;
  }

  public void setLastModifiedDate(File outFile) {
    if (Config.getInstance().isFetchTs()) {
      try {
        Date date = HttpDateParser.parseHttpDate(this.lastModified);
        if (date != null) {
          boolean val = outFile.setLastModified(date.getTime());
          logger.info("Updated last modified: {}", val);
        }
      } catch (Exception e) {
        logger.error("Error setLastModifiedDate", e);
      }
    }
  }

  public void throttle() {
    this.speedLimiter.throttleIfNeeded();
  }

  protected String getOutputFileName() {
    return listener.getOutputFileName(id);
  }
}

File: xdm-core/src/main/java/xdm/core/downloaders/AbstractDownloader.kt
//package xdm.core.downloaders
//
//import xdm.core.Config
//import xdm.core.DownloadProgressListener
//import xdm.core.InteractiveCredentialProvider
//import xdm.core.util.HttpDateParser
//import xdm.core.util.Logger
//import java.io.File
//import java.io.IOException
//import java.nio.file.Files
//import java.nio.file.Path
//import java.nio.file.Paths
//import java.util.concurrent.atomic.AtomicBoolean
//import java.util.concurrent.atomic.AtomicInteger
//import java.util.concurrent.atomic.AtomicLong
//
//abstract class AbstractDownloader protected constructor(
//    protected val id: Long,
//    val downloaderType: DownloaderType,
//    protected val listener: DownloadProgressListener,
//    protected var credentialProvider: InteractiveCredentialProvider
//) {
//    val stopFlag: AtomicBoolean = AtomicBoolean(false)
//    val length: AtomicLong = AtomicLong()
//    var folder: String? = null
//    val finished: AtomicBoolean = AtomicBoolean(false)
//    var maxCount: Int = 8
//    val speedLimiter: SpeedLimiter = SpeedLimiter(this)
//    val downloaded: AtomicLong = AtomicLong(0)
//    val progress: AtomicInteger = AtomicInteger(0)
//    var lastUpdated: Long = 0
//    var lastSaved: Long = 0
//    var assembling: Boolean = false
//    var downloadSpeed: Float = 0f
//    var eta: Long = 0
//    var outputFormat: Int = 0
//    var converting: Boolean = false
//    var convertPrg: Int = 0
//    var lastModified: String? = null
//
//    abstract fun start()
//
//    abstract fun stop()
//
//    abstract fun resume()
//
//    abstract val type: Int
//
//    val size: Long
//        get() = length.get()
//
//    fun getDownloaded(): Long {
//        return downloaded.get()
//    }
//
//    abstract val sourceInfo: SourceInfo?
//
//    abstract val segmentDetails: SegmentDetails?
//
//    @Synchronized
//    fun cleanup() {
//        Logger.info("XDM", "Cleaning up temp files")
//        if (folder == null) {
//            return
//        }
//        try {
//            val folderPath = Paths.get(folder!!)
//            if (!Files.exists(folderPath)) {
//                return
//            }
//            Files.walk(folderPath).sorted(Comparator.reverseOrder()).use { files ->
//                files.forEach { f: Path? ->
//                    try {
//                        Files.delete(f)
//                        Logger.info("XDM", "Successfully delete file : $f")
//                    } catch (e: IOException) {
//                        Logger.error("XDM", "Error deleting temp file: $f ", e)
//                    }
//                }
//            }
//        } catch (ex: Exception) {
//            Logger.error("XDM", "Error cleaning up temp files", ex)
//        }
//    }
//
//    protected val outputFolder: String
//        get() = listener.getOutputFolder(id)
//
//    protected fun getBackupFile(folder: String): File? {
//        val f = File(folder)
//        val files = f.listFiles()
//        if (files == null || files.isEmpty()) return null
//        for (file in files) {
//            if (file.name.endsWith(".bak")) {
//                return file
//            }
//        }
//        return null
//    }
//
//    fun setLastModifiedDate(outFile: File) {
//        if (Config.getInstance().isFetchTs) {
//            try {
//                val date = HttpDateParser.parseHttpDate(this.lastModified)
//                if (date != null) {
//                    val modified = outFile.setLastModified(date.time)
//                    Logger.info("XDM", "Updated last modified: $modified")
//                }
//            } catch (e: Exception) {
//                Logger.error("XDM", "Error setLastModifiedDate", e)
//            }
//        }
//    }
//
//    fun throttle() {
//        speedLimiter.throttleIfNeeded()
//    }
//
//    protected val outputFileName: String
//        get() = listener.getOutputFileName(id)
//}
